compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEAssignNpuDmaBdIds.cpp

```diff
@@ -9,14 +9,165 @@
+    auto logicalObjFifo =
+        dyn_cast_if_present<AMDAIE::LogicalObjFifoOpInterface>(
+            npuDmaOp.getSource().getDefiningOp());
+    if (!logicalObjFifo)
+      return npuDmaOp.emitOpError() << "expected a source logical objectFifo";
+    tiles = logicalObjFifo.getTiles();
+  }
+  if (npuDmaOp.getTarget()) {
+    auto logicalObjFifo =
+        dyn_cast_if_present<AMDAIE::LogicalObjectFifoFromMemrefOp>(
+            npuDmaOp.getTarget().getDefiningOp());
+    if (!logicalObjFifo)
+      return npuDmaOp.emitOpError()
+             << "expected a target `amdaie.logicalobjectfifo.from_memref`";
+    tiles = logicalObjFifo.getTiles();
+  }
+  if (tiles.size() != 1) {
+    if (tiles.empty()) {
+      return npuDmaOp.emitOpError() << "no tiles found";
+    } else {
+      return npuDmaOp.emitOpError()
+             << "operating on multiple tiles is not supported";
+    }
+  }
+  Value tile = tiles[0];
+  if (!shimTileToGeneratorMap.contains(tile))
+    return npuDmaOp.emitOpError()
+           << "no channel BD ID generator found for tile: " << tile;
+
+  auto tileOp = dyn_cast_if_present<AMDAIE::TileOp>(tile.getDefiningOp());
+  if (!tileOp) return npuDmaOp.emitOpError() << "no tile op found";
+  return tileOp;
+};
+
+// Check if the DMA operation is in the innermost loop of controlcode.
+bool isInMostInnerLoop(AMDAIE::NpuDmaCpyNdOp op) {
+  auto parentLoop = op->getParentOfType<scf::ForOp>();
+  if (!parentLoop) return false;
+
+  bool hasNestedLoop = false;
+  parentLoop.walk([&](scf::ForOp nestedLoop) {
+    if (nestedLoop != parentLoop) hasNestedLoop = true;
+  });
+  return !hasNestedLoop;
+}
+
+// Count the number of BD IDs needed per loop iteration,
+// so that we know where to start the BD ID for the next iteration.
+uint32_t countBdIdPerLoopIteration(
+    scf::ForOp loop, AMDAIE::TileOp tileOp,
+    DenseMap<Value, ChannelBdIdGenerator> &shimTileToGeneratorMap) {
+  uint32_t count = 0;
+  loop.walk([&](AMDAIE::NpuDmaCpyNdOp dmaOp) {
+    if (dmaOp.getSource() || dmaOp.getTarget()) {
+      FailureOr<AMDAIE::TileOp> tile =
+          getGeneratorTileOp(dmaOp, shimTileToGeneratorMap);
+      if (succeeded(tile) && *tile == tileOp) count++;
+    }
+  });
+  return count;
+}
+
+FailureOr<AMDAIE::BdIdOp> getBdIdOp(
+    IRRewriter &rewriter, AMDAIE::NpuDmaCpyNdOp &npuDmaOp,
+    DenseMap<Value, ChannelBdIdGenerator> &shimTileToGeneratorMap,
+    DenseMap<TileOp, uint32_t> &tileToBdIdOffsetMap,
+    DenseMap<TileOp, uint32_t> &tileToBdIdSizeMap, uint32_t channel) {
+  FailureOr<AMDAIE::TileOp> tileOp =
+      getGeneratorTileOp(npuDmaOp, shimTileToGeneratorMap);
+  if (failed(tileOp)) return failure();
+
+  ChannelBdIdGenerator &generator = shimTileToGeneratorMap[tileOp->getResult()];
+  std::optional<uint32_t> bdId =
+      generator.getAndAssignBdId(channel, BdIdAssignmentMode::Incremental);
+  if (!bdId) return failure();
+  AMDAIE::BdIdOp bdIdOp;
+  rewriter.setInsertionPoint(npuDmaOp);
+  if (isInMostInnerLoop(npuDmaOp)) {
```

Comment:
Why do you need to check whether this is the innermost loop? Shouldn't this always work?

Reply:
As explained before